site/content/features/gatherers.html

```diff
@@ -0,0 +1,395 @@
+      let n2 = document.createElement('sandbox-source')
+      if (text.match(/void\s+main/) !== null) mainclass = clazz
+      n2.setAttribute('name', clazz + '.java')
+      n2.textContent = text.trimEnd()
+      n.appendChild(n2)
+    }
+    n.setAttribute('version', version)
+    n.setAttribute('mainsource', mainclass + '.java')
+    n.setAttribute('preview', 'true')
+    n.setAttribute('v-cloak', '')
+    e.parentNode.insertBefore(n, e)
+  }
+
+  let sandboxes = document.getElementsByClassName('sandbox')
+  for (let i = sandboxes.length - 1; i >= 0; i--) 
+    sandboxes[i].remove()
+
+  new Vue({ el: '#content' })
+
+  for (const d of document.getElementsByClassName('ace_editor')) {
+    const e = ace.edit(d)
+    e.setOptions({
+      fontSize: "1em"
+    })
+  }
+})
+      //]]>
+    </script>
+      <h2>What's Missing from Streams?</h2>
+      <p>Similar to an iterator, a Java stream yields a sequence of elements. But the stream elements are processed lazily. When you program a stream pipeline, you specify functions that inspect or transform elements:</p>
+<div class='sandbox'>    
+      <pre>void main() {
+    var words = """
+Four score and seven years ago our fathers brought forth on this continent,
+a new nation, conceived in Liberty, and dedicated to the proposition
+that all men are created equal.
+""".split(",\\s+");
+    var longWords = Stream.of(words)
+        .filter(s -&gt; s.length() &gt; 6)
+        .map(String::toLowerCase)
+        .limit(5)
+        .toList();
+    println(longWords);
+}
+</pre></div>
+      <p>The lambda expression <code>s -&gt; s.length() &gt; 6</code> and the <code>toLowerCase</code> method are only called when needed; that is, until five results were collected.</p>
+      <p>The lambda expressions in the calls to <code>filter</code> and <code>map</code> see one element at a time. But sometimes you want to see more than one stream element in order to make a decision</p>
+      <p>Consider the task of eliminating adjacent duplicates. When processing the input <code>The the the quick brown fox jumps over the lazy dog dog</code>, we want to drop the duplicate adjacent <code>the</code> and <code>dog</code>.</p>
+      <p>The stream API has three methods for selectively dropping elements: <code>dropWhile</code>, <code>takeWhile</code>, and <code>filter</code>. The first two don't fit our situation. But we can call <code>filter</code> with a handcrafted function. That function receives the current element and compares it with the predecessor that it has stashed away. </p>
+      <p>Here is a potential implementation:</p>
+      <div class='sandbox'>      
+      <pre>void main() {
+    var sentence = "The the the quick brown fox jumps over the lazy dog dog";
+    Stream&lt;String&gt; words = new Scanner(sentence).tokens();
+    // Note that the lambda could not mutate a variable
+    // String previous;
+    class State { String previous; };
+    var state = new State();
```

Comment:
Not sure whether this is too tricky, but these two lines could also be written as:

```suggestion
    var state = new Object() { String previous; };
```

Reply:
Ok, reading on the named class is better as it is used in the `Gatherer`.